client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -865,6 +911,28 @@ public void close() {
         }
     }
 
+    @Override
+    public CompletableFuture<String> getOrRefeshDelegationTokenFor(String scope, String streamName) {
+        Exceptions.checkNotClosed(closed.get(), this);
```

Comment:
i think we should add `synchronized` to this method to ensure that we are not making concurrent calls to get tokens as that is wasteful.

also, we should cache the recent token, and only make the network call iff some predicate says the token should be refreshed.

Reply:
1. There will be indepedent delegation tokens for different resources. There can be concurrent calls to this method for different resources. Ideally these should not block each other.
2. Caching is an important function. Right now I am not implementing this. Will log an issue for this.

Reply:
Related to the comment of having a different DT per resource, the resource here is the stream right?  Seems the cache belongs within the `ClientFactoryImpl`.

Reply:
Here are some reasons because of which I think complexity of caching of tokens on the client side may not be warranted.
1. Each call to the controller in the context of the resource will refresh the delegationtoken.
2. Delegationtoken exception (including timeout) will enforce closing of the connection to segmentstore for a particular segment. To re-establish a connection we need a new token, which will be retrieved from controller.